File: src/general/hex.cpp
#include "general/hex.h"

#include <cmath>

using namespace thrive;

double Hex::hexSize = DEFAULT_HEX_SIZE;

// void Hex::setHexSize(double newSize) {
//     Hex::hexSize = newSize;
// }

double
    Hex::getHexSize()
{
    return Hex::hexSize;
}

Float3
    Hex::axialToCartesian(double q, double r)
{
    double x = q * Hex::hexSize * 3.0 / 2.0;
    double z = Hex::hexSize * std::sqrt(3) * (r + q / 2.0);
    return Float3(x, 0, z);
}

Int2
    Hex::cartesianToAxial(double x, double y)
{
	// Getting the cube coordinates.
    double cx = x * (2.0 / 3.0) / Hex::hexSize;
    double cy = y / (Hex::hexSize * std::sqrt(3)) - cx / 2.0;
    double cz = -(cx + cy);

    // Rounding the result.
    double rx = round(cx);
    double ry = round(cy);
    double rz = round(cz);

    double xDiff = std::abs(rx - cx);
    double yDiff = std::abs(ry - cy);
    double zDiff = std::abs(rz - cz);

    if(xDiff > yDiff && xDiff > zDiff)
        rx = -(ry + rz);

    else if(yDiff > zDiff)
        ry = -(rx + rz);

	// Returning the axial coordinates.
    return cubeToAxial(rx, ry, rz);
}

Int3
    Hex::axialToCube(double q, double r)
{
    return Int3(q, r, -(q + r));
}

Int2
    Hex::cubeToAxial(double x, double y, double z)
{
    (void)z;
    return Int2(x, y);
}

Int3
    Hex::cubeHexRound(double x, double y, double z)
{
    double rx = round(x);
    double ry = round(y);
    double rz = round(z);

    double xDiff = std::abs(rx - x);
    double yDiff = std::abs(ry - y);
    double zDiff = std::abs(rz - z);

    if(xDiff > yDiff && xDiff > zDiff)
        rx = -(ry + rz);

    else if(yDiff > zDiff)
        ry = -(rx + rz);

    else
        rz = -(ry + rx);

    return Int3(rx, ry, rz);
}

int64_t
    Hex::encodeAxial(double q, double r)
{
    if(std::abs(q) >= ENCODE_AXIAL_OFFSET || std::abs(r) >= ENCODE_AXIAL_OFFSET)
        LEVIATHAN_ASSERT(false, "Coordinates out of range, q and r need to be "
                                "smaller than ENCODE_AXIAL_OFFSET");

    return (q + ENCODE_AXIAL_OFFSET) * ENCODE_AXIAL_SHIFT + r +
           ENCODE_AXIAL_OFFSET;
}

Int2
    Hex::decodeAxial(int64_t s)
{
    int r = (s % ENCODE_AXIAL_SHIFT) - ENCODE_AXIAL_OFFSET;
    int q = (s - r - ENCODE_AXIAL_OFFSET) / ENCODE_AXIAL_SHIFT -
            ENCODE_AXIAL_OFFSET;
    return Int2(q, r);
}

Int2
    Hex::rotateAxial(double q, double r)
{
    return Int2(-r, q + r);
}

Int2
    Hex::rotateAxialNTimes(double q0, double r0, uint32_t n)
{
    Int2 result(q0, r0);

    for(uint32_t i = 0; i < n % 6; i++)
        result = rotateAxial(result);

    return result;
}

Int2
    Hex::flipHorizontally(double q, double r)
{
    return Int2(-q, q + r);
}

// The Vector3 versions of the functions just unpack
// the vector and call the normal functions.
Float3
    Hex::axialToCartesian(const Int2& hex)
{
    return Hex::axialToCartesian(hex.X, hex.Y);
}

Int2
    Hex::cartesianToAxial(const Float3& hex)
{
    return Hex::cartesianToAxial(hex.X, hex.Z);
}

Int3
    Hex::axialToCube(const Int2& hex)
{
    return Hex::axialToCube(hex.X, hex.Y);
}

Int2
    Hex::cubeToAxial(const Int3& hex)
{
    return Hex::cubeToAxial(hex.X, hex.Y, hex.Z);
}

Int3
    Hex::cubeHexRound(const Float3& hex)
{
    return Hex::cubeHexRound(hex.X, hex.Y, hex.Z);
}

int64_t
    Hex::encodeAxial(const Int2& hex)
{
    return Hex::encodeAxial(hex.X, hex.Y);
}

Int2
    Hex::rotateAxial(const Int2& hex)
{
    return Hex::rotateAxial(hex.X, hex.Y);
}

Int2
    Hex::rotateAxialNTimes(const Int2& hex, uint32_t n)
{
    return Hex::rotateAxialNTimes(hex.X, hex.Y, n);
}

Int2
    Hex::flipHorizontally(const Int2& hex)
{
    return Hex::flipHorizontally(hex.X, hex.Y);
}

File: src/general/Hex.cs
using System;
using System.Collections.Generic;
using Godot;

/// <summary>
///   2D axial coordinate pair.
///   As well as some helper functions for converting to cartesian
/// </summary>
public struct Hex : IEquatable<Hex>
{
    /// <summary>
    ///   Maps a hex side to its direct opposite
    /// </summary>
    public static readonly Dictionary<HexSide, HexSide> OppositeHexSide =
        new()
        {
            { HexSide.Top, HexSide.Bottom },
            { HexSide.TopRight, HexSide.BottomLeft },
            { HexSide.BottomRight, HexSide.TopLeft },
            { HexSide.Bottom, HexSide.Top },
            { HexSide.BottomLeft, HexSide.TopRight },
            { HexSide.TopLeft, HexSide.BottomRight },
        };

    /// <summary>
    ///   Each hex has six neighbours, one for each side. This table
    ///   maps the hex side to the coordinate offset of the neighbour
    ///   adjacent to that side.
    /// </summary>
    public static readonly Dictionary<HexSide, Hex> HexNeighbourOffset =
        new()
        {
            { HexSide.Top, new Hex(0, 1) },
            { HexSide.TopRight, new Hex(1, 0) },
            { HexSide.BottomRight, new Hex(1, -1) },
            { HexSide.Bottom, new Hex(0, -1) },
            { HexSide.BottomLeft, new Hex(-1, 0) },
            { HexSide.TopLeft, new Hex(-1, 1) },
        };

    public int Q;
    public int R;

    public Hex(int q, int r)
    {
        Q = q;
        R = r;
    }

    /// <summary>
    ///   Enumeration of the hex sides, clock-wise
    /// </summary>
    public enum HexSide
    {
        /// <summary>
        ///   Directly up
        /// </summary>
        Top = 1,

        /// <summary>
        ///   Up and to the right
        /// </summary>
        TopRight = 2,

        /// <summary>
        ///   Down and to the right
        /// </summary>
        BottomRight = 3,

        /// <summary>
        ///   Directly down
        /// </summary>
        Bottom = 4,

        /// <summary>
        ///   Down and left
        /// </summary>
        BottomLeft = 5,

        /// <summary>
        ///   Up and left
        /// </summary>
        TopLeft = 6,
    }

    public static Hex operator +(Hex a, Hex b)
    {
        return new Hex(a.Q + b.Q, a.R + b.R);
    }

    public static Hex operator -(Hex a, Hex b)
    {
        return new Hex(a.Q - b.Q, a.R - b.R);
    }

    public static Hex operator *(Hex a, int b)
    {
        return new Hex(a.Q * b, a.R * b);
    }

    public static bool operator ==(Hex left, Hex right)
    {
        return left.Equals(right);
    }

    public static bool operator !=(Hex left, Hex right)
    {
        return !(left == right);
    }

    /// <summary>
    ///   Converts axial hex coordinates to cartesian coordinates.
    /// </summary>
    /// <returns>Cartesian coordinates of the hex's center.</returns>
    public static Vector3 AxialToCartesian(Hex hex)
    {
        float x = hex.Q * Constants.DEFAULT_HEX_SIZE * 3.0f / 2.0f;
        float z = Constants.DEFAULT_HEX_SIZE * Mathf.Sqrt(3) * (hex.R + hex.Q / 2.0f);
        return new Vector3(x, 0, z);
    }

    /// <summary>
    ///   Converts cartesian coordinates to axial hex coordinates.
    /// </summary>
    /// <returns>Hex position.</returns>
    public static Hex CartesianToAxial(Vector3 pos)
    {
        // Getting the cube coordinates.
        float cx = pos.x * (2.0f / 3.0f) / Constants.DEFAULT_HEX_SIZE;
        float cy = pos.z / (Constants.DEFAULT_HEX_SIZE * Mathf.Sqrt(3)) - cx / 2.0f;
        float cz = -(cx + cy);

        // Rounding the result.
        float rx = Mathf.Round(cx);
        float ry = Mathf.Round(cy);
        float rz = Mathf.Round(cz);

        float xDiff = Mathf.Abs(rx - cx);
        float yDiff = Mathf.Abs(ry - cy);
        float zDiff = Mathf.Abs(rz - cz);

        if (xDiff > yDiff && xDiff > zDiff)
        {
            rx = -(ry + rz);
        }
        else if (yDiff > zDiff)
        {
            ry = -(rx + rz);
        }

        // Returning the axial coordinates.
        return CubeToAxial(new Int3((int)rx, (int)ry, (int)rz));
    }

    /// <summary>
    ///   Converts axial hex coordinates to coordinates in the cube based hex model
    /// </summary>
    public static Int3 AxialToCube(Hex hex)
    {
        return new Int3(hex.Q, hex.R, -(hex.Q + hex.R));
    }

    /// <summary>
    ///   Converts cube based hex coordinates to axial hex
    ///   coordinates. Basically just seems to discard the z value.
    /// </summary>
    /// <returns>hex coordinates.</returns>
    public static Hex CubeToAxial(Int3 cube)
    {
        return new Hex(cube.x, cube.y);
    }

    /// <summary>
    ///   Correctly rounds fractional hex cube coordinates to the
    ///   correct integer coordinates.
    /// </summary>
    public static Int3 CubeHexRound(Vector3 pos)
    {
        float rx = Mathf.Round(pos.x);
        float ry = Mathf.Round(pos.y);
        float rz = Mathf.Round(pos.z);

        float xDiff = Mathf.Abs(rx - pos.x);
        float yDiff = Mathf.Abs(ry - pos.y);
        float zDiff = Mathf.Abs(rz - pos.z);

        if (xDiff > yDiff && xDiff > zDiff)
        {
            rx = -(ry + rz);
        }
        else if (yDiff > zDiff)
        {
            ry = -(rx + rz);
        }
        else
        {
            rz = -(ry + rx);
        }

        return new Int3((int)rx, (int)ry, (int)rz);
    }

    /// <summary>
    ///   Rotates a hex by 60 degrees about the origin clock-wise.
    /// </summary>
    public static Hex RotateAxial(Hex hex)
    {
        return new Hex(-hex.R, hex.Q + hex.R);
    }

    /// <summary>
    ///   Rotates a hex by (60 * n) degrees about the origin clock-wise.
    /// </summary>
    public static Hex RotateAxialNTimes(Hex original, int n)
    {
        Hex result = original;

        for (int i = 0; i < n % 6; i++)
        {
            result = RotateAxial(result);
        }

        return result;
    }

    /// <summary>
    ///   Makes a hex symmetrical horizontally about the (0,x) axis.
    /// </summary>
    public static Hex FlipHorizontally(Hex hex)
    {
        return new Hex(-hex.Q, hex.Q + hex.R);
    }

    /// <summary>
    ///   Returns the RenderPriority for the hex.
    ///   Between 1 and HEX_RENDER_PRIORITY_DISTANCE^2.
    /// </summary>
    /// <returns>RenderPriority</returns>
    public static int GetRenderPriority(Hex hex)
    {
        return hex.Q.PositiveModulo(Constants.HEX_RENDER_PRIORITY_DISTANCE) * Constants.HEX_RENDER_PRIORITY_DISTANCE
            + hex.R.PositiveModulo(Constants.HEX_RENDER_PRIORITY_DISTANCE) + 1;
    }

    public bool Equals(Hex other)
    {
        return Q == other.Q && R == other.R;
    }

    public override bool Equals(object? obj)
    {
        if (!(obj is Hex hex))
            return false;

        return Equals(hex);
    }

    public override int GetHashCode()
    {
        var hashCode = -1997189103;

        // ReSharper disable NonReadonlyMemberInGetHashCode
        hashCode = (hashCode * -1521134295) + Q.GetHashCode();
        hashCode = (hashCode * -1521134295) + R.GetHashCode();

        // ReSharper restore NonReadonlyMemberInGetHashCode
        return hashCode;
    }

    public override string ToString()
    {
        return Q + ", " + R;
    }
}
